hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieCompactedLogRecordScanner.java

```diff
@@ -63,69 +68,100 @@
-          switch (r.getBlockType()) {
-            case AVRO_DATA_BLOCK:
-              log.info("Reading a data block from file " + s.getPath());
-              // If this is a avro data block, then merge the last block records into the main result
-              merge(records, recordsFromLastBlock);
-              // Load the merged records into recordsFromLastBlock
-              HoodieAvroDataBlock dataBlock = (HoodieAvroDataBlock) r;
-              loadRecordsFromBlock(dataBlock, recordsFromLastBlock);
-              break;
-            case DELETE_BLOCK:
-              log.info("Reading a delete block from file " + s.getPath());
-              // This is a delete block, so lets merge any records from previous data block
-              merge(records, recordsFromLastBlock);
-              // Delete the keys listed as to be deleted
-              HoodieDeleteBlock deleteBlock = (HoodieDeleteBlock) r;
-              Arrays.stream(deleteBlock.getKeysToDelete()).forEach(records::remove);
-              break;
-            case COMMAND_BLOCK:
-              log.info("Reading a command block from file " + s.getPath());
-              // This is a command block - take appropriate action based on the command
-              HoodieCommandBlock commandBlock = (HoodieCommandBlock) r;
-              if (commandBlock.getType() == HoodieCommandBlockTypeEnum.ROLLBACK_PREVIOUS_BLOCK) {
-                log.info("Rolling back the last data block read in " + s.getPath());
-                // rollback the last read data block
-                recordsFromLastBlock.clear();
-              }
-              break;
-            case CORRUPT_BLOCK:
-              log.info("Found a corrupt block in " + s.getPath());
-              // If there is a corrupt block - we will assume that this was the next data block
-              // so merge the last block records (TODO - handle when the corrupted block was a tombstone written partially?)
-              merge(records, recordsFromLastBlock);
-              recordsFromLastBlock.clear();
+        HoodieLogFormatReader reader = new HoodieLogFormatReader(fs, logFile, readerSchema);
+        while(reader.hasNext()) {
+            HoodieLogBlock r = reader.next();
+            String blockInstantTime = r.getMetadata().stream()
+                    .filter(HoodieLogBlockMetadata.I_TIME).findFirst().get().getValue();
+            if(!HoodieTimeline.compareTimestamps(blockInstantTime, this.latestInstantTime,
+                    HoodieTimeline.LESSER_OR_EQUAL)) {
+                //hit a block with instant time greater than should be processed, stop processing further
               break;
+            }
+            switch (r.getBlockType()) {
+              case AVRO_DATA_BLOCK:
+                log.info("Reading a data block from file " + logFile.getPath());
+                // If this is a avro data block, then merge the last block records into the main result
+                merge(records, lastBlocks);
```

Comment:
So idea here (even before this PR) seems to be that we merge the last block, once we read a new block i.e we buffer upto 1 log block worth of additional records?

Reply:
Yes, that's the idea, barring the insert + delete use case explained above. Also, changes to store the `last block` instead of `last block records` as before is to be able to rollback `delete` block.

Reply:
Not fully groking corner cases atm.. Go on for now, sounds mostly correct to me..